### ISSUE TITLE
[GSdx] Auto-switching to GSdx software rendering (manual game fix) with Rogue Galaxy sometimes make PCSX2 freeze

### ISSUE THREAD

**Issue by piorrro33** [PCSX2/pcsx2#1189]
PCSX2 1.5.0-20160215202751 - compiled on Feb 15 2016
Emulation options: safest except MTVU and the game fix enabled
Plugins: http://a.pomf.cat/cfrtwr.png
GSdx settings: http://a.pomf.cat/fslxnj.png
SPU2-X settings: http://a.pomf.cat/tiuswt.png
Last known version to work: I think it never worked.
PC specs: Intel Core i7-4790 @ 3.60 GHz (working with turbo mode at 4 GHz); ASUS GeForce GTX 750 Ti 2GB
Game: Rogue Galaxy PAL SCES-54552

How to reproduce:
0) Enable this fix http://a.pomf.cat/iaakxb.png
1) Go into the Dorgenark, and travel to any planet (works too with any cutscene that toggles the SW mode)
2) Press Start then triangle to skip the cutscene
3) If you're too fast, it won't freeze, too slow, you'll get a black screen since it switched to HW but when you press start it'll be back working since it switches back to SW
4) Enjoy the freeze if you're between too fast and too slow.

Should also crash with D3D9 and OGL, but only tried with D3D11. (if you want me to try, tell me)
What I think is causing this problem, is that it tries to switch to the hardware renderer when you're pausing but since it doesn't have enough time to switch back to software before the cutscene gets skipped, it gets skipped in hardware mode which causes a freeze. I'm not sure it is that, though.

Side note: enabling this fix is MANDATORY (or you're forced to only play with SW), since if you don't, you can get missing text after flashback cutscenes (which are fucked in HW) and other bad stuff.

I hope it was complete enough, but if you need anything else, feel free to ask me.

EDIT: It seems like it doesn't freeze with the OGL renderer, since it doesn't pause the emulation when pressing the pause button, unlike D3D9 or D3D11.

EDIT N°2: If you don't get it to crash the first time, try, try again. It's very very random so sometimes you freeze but thankfully most of the time you don't.

**Comment by MrCK1** [comment]
According to your description, it seems clear that the game is intended to be played in SW mode. 

Manual game fixes are disabled by default for a reason because they often cause more compatibility problems.

Otherwise, it could be a plugin issue related to layer searching. Rogue Galaxy has arbitrary layer sizes; so the search must start from an arbitrary position

**Comment by piorrro33** [comment]
Oh I think I could try doing the same research on the Japanese version 
which is a single-layer DVD and see if it does the same bug. If the bug 
isn't reproduced, that means that the bug is caused by the layer switching.

**Comment by gregory38** [comment]
Is the rendering that broken on latest openGL (with the help of various accurate option).

**Comment by piorrro33** [comment]
@gregory38 Should I try with OGL HW or SW?

**Comment by piorrro33** [comment]
Done a quick test with OGL SW and no problem to report, so far.

**Comment by ramapcsx2** [comment]
Please try it in hardware. Also look out for waterfall graphics that are in some levels. They used to display wrong in hardware / DX but with all the recent improvements in OGL I hope they're fixed now.

**Comment by piorrro33** [comment]
I know the problem with the waterfall and I'm going to check if this issue is still here or not in a minute.

**Comment by piorrro33** [comment]
I found a workaround: enable **Preload Data Frame** in HW hacks.

**Comment by turtleli** [comment]
The freeze issue isn't limited to Rogue galaxy. It seems the main thread sends an event to the SysExecutor thread and waits for that event to complete. It stops processing events while it's doing that. Unfortunately, the SysExecutor thread might have deadlocked.

A similar freeze can also be reproduced by rapidly pressing F9 (Easiest with D3D9/11 renderers, haven't managed to make it freeze when using OGL).

Main thread:
```
[External Code]
[Frames below may be incorrect and/or missing, no symbols loaded for ntdll.dll]
pcsx2-dbg.exe!BaseScopedCoreThread::PostToSysExec(BaseSysExecEvent_ScopedCore * msg) Line 703
pcsx2-dbg.exe!ScopedCoreThreadPause::ScopedCoreThreadPause(BaseSysExecEvent_ScopedCore * abuse_me) Line 747
pcsx2-dbg.exe!DoFmvSwitch() Line 525
pcsx2-dbg.exe!Pcsx2App::LogicalVsync() Line 544
pcsx2-dbg.exe!Pcsx2AppMethodEvent::InvokeEvent() Line 267
pcsx2-dbg.exe!pxActionEvent::_DoInvokeEvent() Line 447
pcsx2-dbg.exe!wxAppWithHelpers::OnInvokeAction(pxActionEvent & evt) Line 658
pcsx2-dbg.exe!Pcsx2App::HandleEvent(wxEvtHandler * handler, void(wxEvtHandler::*)(wxEvent &) func, wxEvent & event) Line 602
*snip*
```

SysExecutorThread
```
[External Code]
[Frames below may be incorrect and/or missing, no symbols loaded for ntdll.dll]
pcsx2-dbg.exe!pthread_mutex_lock(pthread_mutex_t_ * * mutex) Line 268
pcsx2-dbg.exe!SysExecEvent::_DoInvokeEvent() Line 123
pcsx2-dbg.exe!pxEvtQueue::ProcessEvents(std::list<SysExecEvent *,std::allocator<SysExecEvent *> > & list, bool isIdle) Line 218
pcsx2-dbg.exe!pxEvtQueue::ProcessPendingEvents() Line 249
pcsx2-dbg.exe!ExecutorThread::ExecuteTaskInThread() Line 544
pcsx2-dbg.exe!Threading::pxThread::_try_virtual_invoke(void(Threading::pxThread::*)() method) Line 545
pcsx2-dbg.exe!Threading::pxThread::_internal_execute() Line 645
pcsx2-dbg.exe!Threading::pxThread::_internal_callback(void * itsme) Line 685
pcsx2-dbg.exe!ptw32_threadStart(void * vthreadParms) Line 184
 [External Code]	 
```

**Comment by gregory38** [comment]
There is a strange warning on coverity (hang detected). Might worth to give a look.

**Comment by RedDevilus** [comment]
Seems to be fixed in late 2019 tested with a 1.5 version. Probably #3072. Game needed SW mode for new game and load game and that part when u go to different planet i have tested with different renderers. I didn't know that manual gamefix needed to be turned of either because it has been implemented in automatic gamefixes.

TL;DR I played years with manual gamefix Switch to SW mode for FMV , recent addition fixed both start and this issue.
